src/encryption.rs

```diff
@@ -0,0 +1,1021 @@
+pub(crate) enum Mode {
+    Encrypt,
+    Decrypt
+}
+
+impl From<Mode> for openssl::symm::Mode {
+    fn from(v: Mode) -> openssl::symm::Mode {
+        match v {
+            Mode::Encrypt => openssl::symm::Mode::Encrypt,
+            Mode::Decrypt => openssl::symm::Mode::Decrypt
+        }
+    }
+}
+
+impl CryptoScheme {
+    pub(crate) fn key_len(&self) -> usize {
+        self.cipher().key_len()
+    }
+    pub(crate) fn iv_len(&self) -> usize {
+        self.cipher().iv_len().expect("needs iv")
+    }
+    pub(crate) fn tag_len(&self) -> usize {
+        match self {
+            CryptoScheme::Aes256Gcm => AES_GCM_TAG_BYTES,
+            CryptoScheme::Aes128Cbc => 0
+        }
+    }
+    pub(crate) fn cipher(&self) -> Cipher {
+        match self {
+            CryptoScheme::Aes256Gcm => Cipher::aes_256_gcm(),
+            CryptoScheme::Aes128Cbc => Cipher::aes_128_cbc(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct Iv {
+    bytes: Vec<u8>
+}
+
+impl Deref for Iv {
+    type Target = [u8];
+
+    #[inline(always)]
+    fn deref(&self) -> &[u8] {
+        self.bytes.as_slice()
+    }
+}
+
+impl DerefMut for Iv {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut [u8] {
+        self.bytes.as_mut_slice()
+    }
+}
+
+impl Iv {
+    pub(crate) fn from_base64(iv: impl AsRef<str>) -> Result<Iv, base64::DecodeError> {
+        Ok(Iv { bytes: BASE64_STANDARD.decode(iv.as_ref())? })
+    }
+    pub(crate) fn generate(len: usize) -> Iv {
+        let mut bytes = vec![0; len];
+        OsRng.fill_bytes(&mut bytes);
+        Iv { bytes }
+    }
+    #[allow(unused)]
+    pub(crate) fn len(&self) -> usize {
+        self.bytes.len()
+    }
+    pub(crate) fn as_base64(&self) -> String {
+        BASE64_STANDARD.encode(&self.bytes)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct Key {
+    bytes: Vec<u8>
+}
+
+impl Deref for Key {
+    type Target = [u8];
+
+    #[inline(always)]
+    fn deref(&self) -> &[u8] {
+        self.bytes.as_slice()
+    }
+}
+
+impl DerefMut for Key {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut [u8] {
+        self.bytes.as_mut_slice()
+    }
+}
+
+impl Key {
+    pub(crate) fn from_base64(key: impl AsRef<str>) -> Result<Key, base64::DecodeError> {
+        Ok(Key { bytes: BASE64_STANDARD.decode(key.as_ref())? })
+    }
+    pub(crate) fn generate(len: usize) -> Key {
+        let mut bytes = vec![0; len];
+        OsRng.fill_bytes(&mut bytes);
+        Key { bytes }
+    }
+    pub(crate) fn len(&self) -> usize {
+        self.bytes.len()
+    }
+    pub(crate) fn encrypt_aes_128_ecb(self, encryption_key: &Key) -> std::io::Result<EncryptedKey> {
+        let cipher = Cipher::aes_128_ecb();
+        if encryption_key.len() != cipher.key_len() {
+            return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "Invalid key size"));
+        }
+
+        let encrypted_bytes = symm::encrypt(cipher, &encryption_key, None, &self)?;
```

Comment:
Just curious: what does it mean that there is no IV here? Are we not using IVs for keys?

Reply:
I guess because we encrypt keys in ECB which does not take IV.